ktor-network/jvm/src/io/ktor/network/sockets/DatagramSendChannel.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.utils.io.core.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.selects.*
+import kotlinx.coroutines.sync.*
+import java.net.*
+import java.nio.*
+import java.nio.channels.*
+
+internal class DatagramSendChannel(
+    val channel: DatagramChannel,
+    val socket: DatagramSocketImpl
+) : SendChannel<Datagram> {
+    @ExperimentalCoroutinesApi
+    override val isClosedForSend: Boolean
+        get() = socket.isClosed
+
+    @ExperimentalCoroutinesApi
+    override val isFull: Boolean
+        get() = lock.isLocked
+
+    private val lock = Mutex()
+
+    override fun close(cause: Throwable?): Boolean {
+        if (socket.isClosed) {
+            return false
+        }
+
+        socket.close()
+        return true
+    }
+
+
+    override fun offer(element: Datagram): Boolean {
+        if (!lock.tryLock()) return false
+
+        try {
+            val buffer = element.prepareMessage()
+            return channel.send(buffer, element.address) != 0
+        } finally {
+            lock.unlock()
+        }
+    }
+
+    override suspend fun send(element: Datagram) {
+        lock.withLock {
+            val buffer = element.prepareMessage()
+
+            val rc = channel.send(buffer, element.address)
+            if (rc != 0) {
+                socket.interestOp(SelectInterest.WRITE, false)
+                return
+            }
+
+            sendSuspend(buffer, element.address)
```

Comment:
I hope the state machine elimination didn't fail here because of the lambda block `withLock`

Reply:
I'm not sure that it's critical here, but `withLock` is inline, so I hope for the best.